Add flatMap method for collapsing streams into one

sleeplesslord opened the pull request:
Motivation:
`StreamMessage` has `map`, but it is inconvenient to use it when the mapping function itself returns a `StreamMessage`. This PR adds `flatMap` which allows applying a function that returns a `StreamMessage` without ending up with nested `StreamMessage`s.

Modifications:
- Add `FlatMapStreamMessage`, which uses a function that returns a `StreamMessage` to modify the stream.
- Add `StreamMessage.flatMap()` to allow easily creating a `FlatMapStreamMessage` from an existing stream.

Result:
- Closes #3966 
- You can now use `StreamMessage.flatMap` to modify a stream using a function that returns a `StreamMessage`.

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,361 @@
+                return;
+            }
+
+            final long available = getAvailableBufferSpace();
```
If `available` is `Long.MAX_VALUE`, we don't need backpressure with buffering.

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,361 @@
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            upstream.request(maxConcurrency - sourceSubscriptions.size());
```
It should never happen

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,361 @@
+            }
+
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this));
```
Is it OK to just pass executor here? I don't think there are any options that need to be applied

trustin replied:
@ikhoon Gentle ping

sleeplesslord replied:
I tried passing the executor here and I'm running into an issue I'm not sure how to address. 

This code calls `subscribe`, which calls `onSubscribe` on `FlatMapSubscriber`. That will call `subscribeChild` on `FlatMapAggregatingSubscriber`, which will add the subscriber to `sourceSubscriptions` and then call `request` on `FlatMapSubscriber`. Then, if `newStreamMessage` is already completed it will call `onComplete`, which leads to the subscription being removed from `sourceSubscriptions`. 

Essentially, in one call chain we go from `onSubscribe` to `sub.request(1)` all the way to `onComplete`, but since that removes the subscription from `sourceSubscriptions` I get a `ConcurrentModificationException` since the call to `sub.request(1)` is done in a `forEach`. 

I guess I should just collect the subs into a new collection before calling `forEach` to avoid that? I'm not sure if there's a better approach

minwoox replied:
I see. How about always using `executor` to remove the child?
`executor.execute(() -> handleCompleteChild(child));`

minwoox replied:
If that works, we can remove all `eventLoop.inEventLoop()` check because we only use the executor.

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,361 @@
+                return Long.MAX_VALUE;
+            }
+
+            final Optional<Long> requested = sourceSubscriptions.stream().map(FlatMapSubscriber::getRequested)
+                                                                .reduce(LongMath::saturatedAdd);
```
Will that handle overflowing in case `getRequested` returns `Long.MAX_VALUE`?

sleeplesslord commented:
Hey, sorry for letting this PR go stale for so long, I was addressing the comments but then I lost track of things 😅 I think I've covered most of the feedback now

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,392 @@
+                                                                               .filter(sub -> sub.getRequested()
+                                                                                              == 0)
+                                                                               .limit(available)
+                                                                               .collect(Collectors.toList());
```
nit:
```suggestion
                                                                               .collect(toImmutableList());
```
Yeah, creating a new list also works. 😄 
So, Could you remove `if (executor.inEventLoop()) {..} else {...}` from this class where possible? 😄

sleeplesslord replied:
Is that because the `FlatMapSubscriber` will run with the same executor, so we know that `completeChild` etc will always run in the same executor?

minwoox replied:
Yes exactly.  😄 
All `onX` methods in `FlatMapAggregatingSubscriber` and `FlatMapSubscriber` are now called by the `executor` so we don't have to check it. 😄

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,368 @@
+
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this), executor);
+            pendingSubscriptions++;
```
I think `pendingSubscriptions` can be `-1`.
Let's do this before subscribing.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,368 @@
+        }
+
+        private void cancelSourceSubscriptions() {
+            sourceSubscriptions.forEach(FlatMapSubscriber::cancel);
```
This isn't actually subscription but subscriber.
How about `childSubscribers`?

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,368 @@
+            sourceSubscriptions.forEach(FlatMapSubscriber::cancel);
+        }
+
+        void subscribeChild(FlatMapSubscriber<T, U> child) {
```
nit: childSubscribed?

minwoox commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,368 @@
+
+            if (available == Long.MAX_VALUE) {
+                sourceSubscriptions.forEach(sub -> sub.request(Long.MAX_VALUE));
+            }
```
```suggestion
                return;
            }
```

minwoox reviewed (commented):
Thanks! Left some more comments. 😉

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,378 @@
+                return;
+            }
+            canceled = true;
+
```
I guess whatever is in the `buffer` needs to be cleaned up/closed using `StreamMessageUtil.closeOrAbort` and also `whenComplete` needs to be completed exceptionally.

Same goes for when `FlatMapAggregatingSubscriber#cancel` is called.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,378 @@
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            flush();
+
+            final long toRequest = maxConcurrency - childSubscribers.size();
+            if (toRequest > 0) {
+                upstream.request(toRequest);
+            }
```
It seems like we should also request upstream once a subscriber completes and is removed from `childSubscribers`.
e.g. 3 child subscribers with 2 maxConcurrency may result in the third child not processing.

```
final StreamMessage<Integer> streamMessage = StreamMessage.of(10, 20, 30);
final Function<Integer, StreamMessage<Integer>> function =
        num -> StreamMessage.of(Flux.range(num, 3).delayElements(Duration.ofSeconds(1)));
final StreamMessage<Integer> mappedStream = streamMessage.flatMap(function, 2);

StepVerifier.create(mappedStream)
            .thenRequest(100L)
            .thenConsumeWhile(num -> {
                System.out.println("num: " + num);
                return true;
            })
            .verifyComplete();
```

jrhee17 replied:
To give a more simple example of the deadlock example I mentioned earlier:
A child stream may not emit any elements and complete the its stream.
In this case, although there are outstanding requests it may be possible that the stream doesn't emit any more elements and a deadlock occurs.

Related to my comment above, I wonder if we want to just request upstream once and be over with it.

```suggestion
            if (!initialized) {
                initialized = true;
                upstream.request(maxConcurrency);
            }
```

and request upstream each time a child completes to maintain `maxConcurrency` count of child streams.


Also as it stands, since child streams are added when `onSubscribe` is invoked it is possible that >3 child streams are subscribed to although `maxConcurrency == 3`

sleeplesslord replied:
Good point, I think that would simplify this part a lot.

>Also as it stands, since child streams are added when onSubscribe is invoked it is possible that >3 child streams are subscribed to although maxConcurrency == 3

I don't understand this part. If you mean `FlatMapSubscriber.onSubscribe`, it does add more child subscribers but the number of `FlatMapSubscriber.onSubscribe` calls is limited by the number of elements requested from the upstream, which is in turn limited by `maxConcurrency`.

jrhee17 replied:
I thought the following scenario was possible (let me know if I'm misunderstanding)

Assuming a `maxConcurrency == 3`
1. FlatMapAggregatingSubscriber.request(3)
2. FlatMapAggregatingSubscriber.request(3) 
4. FlatMapAggregatingSubscriber.onSubscribe() is invoked for 6 elements
5. childSubscribers.size() == 6

sleeplesslord replied:
Ah, you're right 😅 If multiple requests are made before any of the inner subscriptions are created it will make multiple requests to upstream. This is neatly fixed by the suggestion you made with an initialization flag 👍

jrhee17 commented:
Looks good so far 👍 

I just wanted to confirm my understanding on the concept of `maxConcurrency`.
I understood `maxConcurrency` to be the maximum number of child `StreamMessage`s a `FlatMapStreamMessage` is receiving data from.

For instance, assuming `maxConcurrency == 2`.

1. When `FlatMapAggregatingSubscriber#request` is first called, `FlatMapAggregatingSubscriber#upstream.request(maxConcurrency)` regardless of the number of elements requested from the downstream subscription.
```
        private void handleRequest(long n) {
            ...
            if (!initialized) {
                initialized = true;
                upstream.request(maxConcurrency);
            }
            ...
        }

        void completeChild(FlatMapSubscriber<T, U> child) {
        ...
            // to compensate for the completed child which invoked this callback
            if (upstream hasn't completed) {
              upstream.request(1)
            }
        ...
        }
```

2. When a `request(n)` from downstream is made, first try to drain the oustanding queue. Afterwards, request the remaining amount to each child subscriber.

e.g.
if `queueSize == 3` and `FlatMapAggregatingSubscriber.request(n)`, then request to child subscribers such that all `FlatMapSubscriber.requested == n - 3`


I thought by decoupling the concept (number of child streams vs. number of requested elements), the code can be a little cleaner and more deterministic.
Let me know if this doesn't make sense 🙏

sleeplesslord commented:
@jrhee17 
The way it currently works, `maxConcurrency` is, like you said, the maximum number of child subscribers at any one time. It also limits the buffer size - currently the allowed buffer is the same as `maxConcurrency`, so any one subscriber can store one value in the buffer. My thought was to always request `1` from each subscriber as available and store the overflow in the buffer. So, if `maxConcurrency == 3` and downstream requests `1` it would make three child subscribers, request `1` from each, then store two values in the buffer.

>The total number of ongoing requests to child streams would be (n-3 * maxConcurrency) in this case, and overflows would be stored in the queue.

The total number of ongoing requests should be equal to `maxConcurrency - buffer.size()`. Currently the logic is ignoring the number of requests from downstream, so it will always limit itself that way even if downstream's request is large.

jrhee17 commented:
> The total number of ongoing requests should be equal to maxConcurrency - buffer.size()

I see. I think this is what was confusing me 😅 

Assuming an initial state where `maxConcurrency == 3` and `FlatMapStreamMessage.request(1)`, would this mean there is an ongoing request for one subscriber only? (i.e. `child1`?)

I was thinking if `child3` can produce elements faster than `child1`, then it would be more advantageous to eagerly request to all children.

e.g.
- child1.request(1)
- child2.request(1)
- child3.request(1)

and then queue overflowing elements that weren't requested.

sleeplesslord commented:
>Assuming an initial state where maxConcurrency == 3 and FlatMapStreamMessage.request(1), would this mean there is an ongoing request for one subscriber only? (i.e. child1?)

In that case there would be a request for each subscriber. At the start the buffer is empty, so you would get `maxConcurrency - buffer.size() -> 3 - 0 = 3` total requests. Request-per-child is always 1, so 3 total requests means every child gets one request.

jrhee17 commented:
> In that case there would be a request for each subscriber. At the start the buffer is empty, so you would get maxConcurrency - buffer.size() -> 3 - 0 = 3 total requests. Request-per-child is always 1, so 3 total requests means every child gets one request.

I see, I think I understand the intention now 😅  
So `maxConcurrency` is effectively the number of child streams that should be processed at a given time.
Each child stream is requested 1 at maximum, so the overall ongoing requests is bound to `maxConcurrency`.

Thanks for the explanation 👍

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,378 @@
+        }
+
+        @Override
+        public void onError(Throwable cause) {
```
We should also call `completionFuture#completeExceptionally(cause)` in this block

`StreamMessage#whenComplete` is an often used callback in this codebase, so it's important that this future is completed

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,378 @@
+        void completeChild(FlatMapSubscriber<T, U> child) {
+            childSubscribers.remove(child);
+
+            if (childSubscribers.isEmpty() && pendingSubscriptions == 0 && completing) {
```
Is it possible that the queue is non-empty even though this condition is true?
It seems like we should also check if the queue is drained before checking this condition. (ditto for the other condition at `onComplete`)

I think we may also need to check this condition at `handleRequest` as well.

```suggestion
            if (childSubscribers.isEmpty() && pendingSubscriptions == 0 && completing && buffer.empty()) {
```

jrhee17 reviewed (commented):
Left some more comments 🙏 
I think once my comments are addressed this PR should be in a good place for merging.

Let me know if any of my comments doesn't make sense 😄

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,400 @@
+        }
+
+        private void cancelChildSubscribersAndBuffer() {
+            buffer.forEach(StreamMessageUtil::closeOrAbort);
```
Should we also clear the buffer so that closed messages aren't passed by mistake?
```
buffer.clear();
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,400 @@
+
+        @Override
+        public void cancel() {
+            upstream.cancel();
```
Question) Do we also want to call `cancel` here in case this method is called multiple times?
Also, is there no need to reschedule to the event loop similar to `request(n)`? (this is called by the subscriber which isn't guaranteed to call from the `executor`)

```suggestion
            if (canceled) {
                return;
            }
            canceled = true;
            upstream.cancel();
```

jrhee17 reviewed (commented):
I think this PR is almost done! Left a few more minor comments/questions 🙏

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,400 @@
+            if (requestedByDownstream > 0) {
+                publishDownstream(value);
+            } else {
+                buffer.add(value);
```
Ah, for some reason I was remembering this completely incorrectly and thought it was never OK for `onNext` to be called after `cancel`. I think the cleanest way to do this is to have the child subscriber `closeOrAbort` anything it receives after it's cancelled

jrhee17 reviewed (commented):
Almost there 🙏

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,424 @@
+
+        void childSubscribed(FlatMapSubscriber<T, U> child) {
+            pendingSubscriptions--;
+            childSubscribers.add(child);
```
I'm not sure I understand in which scenario a duplicate child can occur. Could you explain it a bit more?

sleeplesslord commented on core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java:
```diff
@@ -0,0 +1,424 @@
+            final Optional<Long> requested = childSubscribers.stream().map(FlatMapSubscriber::getRequested)
+                                                             .reduce(LongMath::saturatedAdd);
+
+            return maxConcurrency - requested.orElse(0L) - buffer.size();
```
I think you're right, but I'm trying to strike a balance between how much buffering is done between the requests. In the example you gave child2 and child3 will be idle, but at the same time there will be two values in the buffer. Once the downstream requests those 2 values all the children will become active again since the buffer is empty.

If each child subscriber has a fixed target the buffer size will become a lot larger, e.g if `maxConcurrency == 1000`, `requestedByDownstream == 1000` the buffer could contain a million values.

jrhee17 reviewed (approved):
I've added a small commit which:

- Addresses a `ConcurrentModificationException` thrown when the `forEach` call inside `#requestAllAvailable` calls `childSubscribers#remove` in `#completeChild`. `childSubscribers` is copied to mitigate this issue.
- `FlatMapStreamMessage` could be closed, whereas the inner `FlatMapSubscriber` may not be closed. This is possible if `abort` is called before `FlatMapSubscriber#onSubscribe` is invoked. I've changed so that `FlatMapSubscriber` is added to `childSubscribers` immediately.
- Miscellaneous possible NPEs were dealt with
- Flaky tests were dealt with

Let me know if any changes don't make sense.

Lastly, sorry this PR took so long 😅     Looks good to me for merging 👍

minwoox commented:
Thanks! I also put some miscellaneous changes.

minwoox reviewed (approved):
Thanks a lot, @sleeplesslord and @jrhee17!

ikhoon reviewed (approved):
Looks great! 🚀 💯

sleeplesslord commented:
Thanks for the fixes and the merge, happy to see this make it in :pray:
